Firestore/core/src/firebase/firestore/model/field_path.cc

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/model/field_path.h"
+
+#include <algorithm>
+#include <cctype>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_replace.h"
+#include "absl/strings/str_split.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+namespace {
+
+// TODO(varconst): move to C++ equivalent of FSTDocumentKey.{h,cc}
+const char* const kDocumentKeyPath = "__name__";
+
+/**
+ * True if the string could be used as a segment in a field path without
+ * escaping.
+ */
+bool IsValidIdentifier(const std::string& segment) {
+  if (segment.empty()) {
+    return false;
+  }
+  if (segment.front() != '_' && !std::isalpha(segment.front())) {
+    return false;
+  }
+  if (std::any_of(segment.begin(), segment.end(), [](const unsigned char c) {
+        return c != '_' && !std::isalnum(c);
+      })) {
+    return false;
+  }
+
+  return true;
+}
+
+std::string EscapedSegment(const std::string& segment) {
+  auto escaped = absl::StrReplaceAll(segment, {{"\\", "\\\\"}, {"`", "\\`"}});
+  const bool needs_escaping = !IsValidIdentifier(escaped);
+  if (needs_escaping) {
+    escaped.insert(escaped.begin(), '`');
+    escaped.push_back('`');
+  }
+  return escaped;
+}
+
+}  // namespace
+
+FieldPath FieldPath::ParseServerFormat(const absl::string_view path) {
+  // TODO(b/37244157): Once we move to v1beta1, we should make this more
+  // strict. Right now, it allows non-identifier path components, even if they
+  // aren't escaped. Technically, this will mangle paths with backticks in
+  // them used in v1alpha1, but that's fine.
+
+  SegmentsT segments;
+  std::string segment;
+  segment.reserve(path.size());
+
+  // string_view doesn't have a c_str() method, because it might not be
+  // null-terminated. Assertions expect C strings, so construct std::string on
+  // the fly, so that c_str() might be called on it.
+  const auto to_string = [](const absl::string_view view) {
+    return std::string{view.data(), view.data() + view.size()};
+  };
+  const auto finish_segment = [&segments, &segment, &path, &to_string] {
+    FIREBASE_ASSERT_MESSAGE(
```

Comment:
In this and some other instances, Objective C throws, so I'm a little concerned that in C++ it's unrecoverable (OTOH, I don't know if any code in Objective C actually catches those throws).

Reply:
Thanks for the explanation.